Queries fail while using native hive formats

Since version `409`  (with https://github.com/trinodb/trino/pull/15918) queries fail on `java.lang.IllegalArgumentException: Compressed files are not splittable` while using  configuration `{$format}.native-reader.enabled=true`

Issue happens when reading external hive tables and only during the time that the underline files are updated.

Issue doesn't happen when the underlined files are not updated.
It also doesn’t happen when
`native-reader.enabled=false`.

External table for example:
```
CREATE TABLE hive.temp.events (
   event_name varchar,
   event_at varchar,
   env varchar,
   payload varchar
)
WITH (
   csv_escape = U&'\0000',
   csv_separator = ',',
   external_location = 's3a://bucket-name/events/last_hour',
   format = 'CSV'
)
```

Files in `s3://bucket-name/events/last_hour`
```
-- Each one minute, a relevant file is replaced. 

s3://bucket-name/events/last_hour/00.gz
s3://bucket-name/events/last_hour/01.gz
s3://bucket-name/events/last_hour/02.gz
.
.
.
s3://bucket-name/events/last_hour/59.gz
```


Query using `events` table:
```
SELECT count(*) 
FROM events
WHERE event_name = 'some_event'
AND json_extract_scalar(payload, '$.status') = 'failed'
```

When `csv.native-reader.enabled=false` no issue happen 

When `csv.native-reader.enabled=true` (the default) queries fail from time to time:

```
io.trino.spi.TrinoException: Error opening Hive split s3a://bucket-name/events/last_hour/59.gz (offset=0, length=18736): Compressed files are not splittable
	at io.trino.plugin.hive.line.LinePageSourceFactory.createPageSource(LinePageSourceFactory.java:191)
	at io.trino.plugin.hive.HivePageSourceProvider.createHivePageSource(HivePageSourceProvider.java:216)
	at io.trino.plugin.hive.HivePageSourceProvider.createPageSource(HivePageSourceProvider.java:154)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:49)
	at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:62)
	at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:266)
	at io.trino.operator.ScanFilterAndProjectOperator$SplitToPages.process(ScanFilterAndProjectOperator.java:194)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:360)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:347)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:347)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:262)
	at io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:241)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:262)
	at io.trino.operator.WorkProcessorUtils.lambda$finishWhen$3(WorkProcessorUtils.java:256)
	at io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:413)
	at io.trino.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:146)
	at io.trino.operator.Driver.processInternal(Driver.java:394)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:297)
	at io.trino.operator.Driver.tryWithLock(Driver.java:689)
	at io.trino.operator.Driver.process(Driver.java:289)
	at io.trino.operator.Driver.processForDuration(Driver.java:260)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:773)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:165)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:523)
	at io.trino.$gen.Trino_409____20230312_083525_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IllegalArgumentException: Compressed files are not splittable
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:145)
	at io.trino.hive.formats.line.text.TextLineReaderFactory.createLineReader(TextLineReaderFactory.java:71)
	at io.trino.plugin.hive.line.LinePageSourceFactory.createPageSource(LinePageSourceFactory.java:179)
	... 31 more
```

cc: @dain @electrum

--- a/lib/trino-hive-formats/src/main/java/io/trino/hive/formats/compression/CompressionKind.java
+++ b/lib/trino-hive-formats/src/main/java/io/trino/hive/formats/compression/CompressionKind.java
@@ -95,9 +95,12 @@ public enum CompressionKind
             .filter(codec -> codec.fileExtension != null)
             .collect(toImmutableMap(codec -> codec.fileExtension, Function.identity()));
 
-    public static Optional<Codec> createCodecFromExtension(String extension)
+    public static Optional<CompressionKind> forFile(String fileName)
     {
-        return Optional.ofNullable(CODECS_BY_FILE_EXTENSION.get(extension))
-                .map(CompressionKind::createCodec);
+        int position = fileName.lastIndexOf('.');
+        if (position < 0) {
+            return Optional.empty();
+        }
+        return Optional.ofNullable(CODECS_BY_FILE_EXTENSION.get(fileName.substring(position)));
     }
 }

--- a/lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/text/TextLineReaderFactory.java
+++ b/lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/text/TextLineReaderFactory.java
@@ -65,8 +65,8 @@ public class TextLineReaderFactory
     {
         InputStream inputStream = inputFile.newStream();
         try {
-            Optional<Codec> codec = getExtension(inputFile.location())
-                    .flatMap(CompressionKind::createCodecFromExtension);
+            Optional<Codec> codec = CompressionKind.forFile(inputFile.location())
+                    .map(CompressionKind::createCodec);
             if (codec.isPresent()) {
                 checkArgument(start == 0, "Compressed files are not splittable");
                 // for compressed input, we do not know the length of the uncompressed text
@@ -102,14 +102,5 @@ public class TextLineReaderFactory
                 return;
             }
         }
-    }
-
-    private static Optional<String> getExtension(String location)
-    {
-        int position = location.lastIndexOf('.');
-        if (position < 0) {
-            return Optional.empty();
-        }
-        return Optional.of(location.substring(position));
     }
 }

--- a/lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/text/TextLineReaderFactory.java
+++ b/lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/text/TextLineReaderFactory.java
@@ -68,7 +68,7 @@ public class TextLineReaderFactory
             Optional<Codec> codec = getExtension(inputFile.location())
                     .flatMap(CompressionKind::createCodecFromExtension);
             if (codec.isPresent()) {
-                checkArgument(start == 0 && length == inputFile.length(), "Compressed files are not splittable");
+                checkArgument(start == 0, "Compressed files are not splittable");
                 // for compressed input, we do not know the length of the uncompressed text
                 length = Long.MAX_VALUE;
                 inputStream = codec.get().createStreamDecompressor(inputStream);

--- a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/line/LinePageSourceFactory.java
+++ b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/line/LinePageSourceFactory.java
@@ -117,12 +117,12 @@ public abstract class LinePageSourceFactory
 
         // get header and footer count
         int headerCount = getHeaderCount(schema);
-        if (headerCount > 0) {
-            checkArgument(estimatedFileSize == start + length, "Header not supported for a split file");
+        if (headerCount > 1) {
+            checkArgument(start == 0, "Multiple header rows are not supported for a split file");
         }
         int footerCount = getFooterCount(schema);
         if (footerCount > 0) {
-            checkArgument(estimatedFileSize == start + length, "Footer not supported for a split file");
+            checkArgument(start == 0, "Footer not supported for a split file");
         }
 
         // setup projected columns

--- a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java
+++ b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java
@@ -27,6 +27,7 @@ import io.airlift.slice.Slice;
 import io.airlift.slice.SliceUtf8;
 import io.airlift.slice.Slices;
 import io.trino.hadoop.TextLineLengthLimitExceededException;
+import io.trino.hive.formats.compression.CompressionKind;
 import io.trino.orc.OrcWriterOptions;
 import io.trino.plugin.hive.HiveColumnHandle;
 import io.trino.plugin.hive.HivePartitionKey;
@@ -415,9 +416,16 @@ public final class HiveUtil
 
     public static boolean isSplittable(InputFormat<?, ?> inputFormat, FileSystem fileSystem, Path path)
     {
-        // ORC uses a custom InputFormat but is always splittable
-        if (inputFormat.getClass().getSimpleName().equals("OrcInputFormat")) {
-            return true;
+        // TODO move this to HiveStorageFormat when Hadoop library is removed
+        switch (inputFormat.getClass().getSimpleName()) {
+            case "OrcInputFormat", "MapredParquetInputFormat", "AvroContainerInputFormat", "RCFileInputFormat", "SequenceFileInputFormat" -> {
+                // These formats have splitting built into the format
+                return true;
+            }
+            case "TextInputFormat" -> {
+                // Only uncompressed text input format is splittable
+                return CompressionKind.forFile(path.getName()).isEmpty();
+            }
         }
 
         // use reflection to get isSplittable method on FileInputFormat

